File: app/src/main/java/net/theluckycoder/scriptcraft/component/CodeEditText.java
package net.theluckycoder.scriptcraft.component;


import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.os.Handler;
import android.support.v4.content.ContextCompat;
import android.support.v7.widget.AppCompatEditText;
import android.text.Editable;
import android.text.InputFilter;
import android.text.Layout;
import android.text.Spannable;
import android.text.SpannableStringBuilder;
import android.text.Spanned;
import android.text.TextWatcher;
import android.text.style.BackgroundColorSpan;
import android.text.style.ForegroundColorSpan;
import android.util.AttributeSet;
import android.util.Log;
import android.util.TypedValue;
import android.view.KeyEvent;
import android.view.ViewTreeObserver;

import net.theluckycoder.scriptcraft.R;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

import static android.preference.PreferenceManager.getDefaultSharedPreferences;

public class CodeEditText extends AppCompatEditText {

    private final Pattern PATTERN_NUMBERS = Pattern.compile("\\b(\\d*[.]?\\d+)\\b");
    private final Pattern PATTERN_CLASSES = Pattern.compile(
            "^[\t ]*(Object|Function|Boolean|Symbol|Error|EvalError|InternalError|" +
                    "RangeError|ReferenceError|SyntaxError|TypeError|URIError|" +
                    "Number|Math|Date|String|RegExp|Map|Set|WeakMap|WeakSet|" +
                    "Array|ArrayBuffer|DataView|JSON|Promise|Generator|GeneratorFunction" +
                    "Reflect|Proxy|Intl)\\b",
            Pattern.MULTILINE);
    private final Pattern PATTERN_KEYWORDS = Pattern.compile(
            "\\b(break|case|catch|class|const|continue|debugger|default|delete|do|yield|" +
                    "else|export|extends|finally|for|function|if|import|in|instanceof|" +
                    "new|return|super|switch|this|throw|try|typeof|var|void|while|with|" +
                    "null|true|false)\\b");
    private final Pattern PATTERN_COMMENTS = Pattern.compile("/\\*(?:.|[\\n\\r])*?\\*/|//.*");
    private Context context;
    private final transient Paint paint = new Paint();
    private Layout layout;
    private final Handler updateHandler = new Handler();
    private OnTextChangedListener onTextChangedListener;
    private final int updateDelay = 1000;
    private boolean modified = true;
    private int colorNumber, colorKeyword, colorBuiltin, colorComment, colorString;
    private final Runnable updateRunnable =
            new Runnable() {
                @Override
                public void run() {
                    Editable e = getText();

                    if (onTextChangedListener != null)
                        onTextChangedListener.onTextChanged(
                                e.toString());

                    highlightWithoutChange(e);
                }
            };
    private BackPressedListener mOnImeBack;

    public CodeEditText(Context context) {
        super(context);

        init(context);
    }

    public CodeEditText(Context context, AttributeSet attrs) {
        super(context, attrs);

        init(context);
    }

    private static void clearSpans(Editable e) {
        // remove foreground color spans
        {
            ForegroundColorSpan spans[] = e.getSpans(0, e.length(), ForegroundColorSpan.class);

            for (int n = spans.length; n-- > 0; )
                e.removeSpan(spans[n]);
        }

        // remove background color spans
        {
            BackgroundColorSpan spans[] = e.getSpans(0, e.length(), BackgroundColorSpan.class);

            for (int n = spans.length; n-- > 0; )
                e.removeSpan(spans[n]);
        }
    }

    public void setTextHighlighted(CharSequence text) {
        if (text == null)
            text = "";

        cancelUpdate();

        modified = false;
        setText(highlight(new SpannableStringBuilder(text)));
        modified = true;

        if (onTextChangedListener != null)
            onTextChangedListener.onTextChanged(text.toString());
    }

    private void init(Context context) {
        setHorizontallyScrolling(true);

        setFilters(new InputFilter[]{
                new InputFilter() {
                    @Override
                    public CharSequence filter(CharSequence source, int start, int end, Spanned dest, int dStart, int dEnd) {
                        if (modified &&
                                end - start == 1 &&
                                start < source.length() &&
                                dStart < dest.length()) {
                            char c = source.charAt(start);

                            if (c == '\n')
                                return autoIndent(source, dest, dStart, dEnd);
                        }

                        return source;
                    }
                }});

        addTextChangedListener(
                new TextWatcher() {

                    @Override
                    public void onTextChanged(CharSequence s, int start, int before, int count) {

                    }

                    @Override
                    public void beforeTextChanged(CharSequence s, int start, int count, int after) {

                    }

                    @Override
                    public void afterTextChanged(Editable e) {
                        cancelUpdate();

                        if (!modified)
                            return;

                        updateHandler.postDelayed(updateRunnable, updateDelay);
                    }
                });

        setSyntaxColors(context);
        this.context = context;
        Paint bgPaint = new Paint();
        bgPaint.setStyle(Paint.Style.FILL);
        bgPaint.setColor(Color.parseColor("#eeeeee"));

        paint.setStyle(Paint.Style.FILL);
        paint.setAntiAlias(true);
        paint.setColor(Color.parseColor("#bbbbbb"));
        paint.setTextSize(getPixels(Integer.parseInt(getDefaultSharedPreferences(context).getString("font_size", "15"))));
        getViewTreeObserver().addOnGlobalLayoutListener(new ViewTreeObserver.OnGlobalLayoutListener() {
            @Override
            public void onGlobalLayout() {
                layout = getLayout();
            }
        });
    }

    private void setSyntaxColors(Context context) {
        colorNumber = ContextCompat.getColor(context, R.color.syntax_number);
        colorKeyword = ContextCompat.getColor(context, R.color.syntax_keyword);
        colorBuiltin = ContextCompat.getColor(context, R.color.syntax_class);
        colorComment = ContextCompat.getColor(context, R.color.syntax_comment);
        colorString = ContextCompat.getColor(context, R.color.syntax_string);
    }

    private void cancelUpdate() {
        updateHandler.removeCallbacks(updateRunnable);
    }

    private void highlightWithoutChange(Editable e) {
        modified = false;
        highlight(e);
        modified = true;
    }

    private Editable highlight(Editable editable) {
        try {
            // don't use e.clearSpans() because it will
            // remove too much
            clearSpans(editable);

            if (editable.length() == 0)
                return editable;

            for (Matcher m = PATTERN_NUMBERS.matcher(editable); m.find(); )
                editable.setSpan(new ForegroundColorSpan(colorNumber), m.start(), m.end(), Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);

            for (Matcher m = PATTERN_CLASSES.matcher(editable); m.find(); )
                editable.setSpan(new ForegroundColorSpan(colorBuiltin), m.start(), m.end(), Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);

            for (Matcher m = PATTERN_KEYWORDS.matcher(editable); m.find(); )
                editable.setSpan(new ForegroundColorSpan(colorKeyword), m.start(), m.end(), Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);

            for(Matcher m = Pattern.compile("\\$\\w+").matcher(editable); m.find(); ) {
                editable.setSpan(new ForegroundColorSpan(colorKeyword), m.start(), m.end(), Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
            }

            for(Matcher m = Pattern.compile("\"(.*?)\"|'(.*?)'").matcher(editable); m.find(); ) {
                ForegroundColorSpan spans[] = editable.getSpans(m.start(), m.end(), ForegroundColorSpan.class);
                for(ForegroundColorSpan span : spans)
                    editable.removeSpan(span);
                editable.setSpan(new ForegroundColorSpan(colorString), m.start(), m.end(), Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
            }

            for (Matcher m = PATTERN_COMMENTS.matcher(editable); m.find(); ) {
                ForegroundColorSpan spans[] = editable.getSpans(m.start(), m.end(), ForegroundColorSpan.class);
                for(ForegroundColorSpan span : spans)
                    editable.removeSpan(span);
                editable.setSpan(new ForegroundColorSpan(colorComment), m.start(), m.end(), Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
            }
        } catch (IllegalStateException e) {
            Log.e("IllegalStateException", e.getMessage(), e);
            // raised by Matcher.start()/.end() when
            // no successful match has been made what
            // shouldn't ever happen because of find()
        }

        return editable;
    }

    private CharSequence autoIndent(CharSequence source, Spanned dest, int dStart, int dEnd) {
        String indent = "";
        int iStart = dStart - 1;

        // find start of this line
        boolean dataBefore = false;
        int pt = 0;

        for (; iStart > -1; --iStart) {
            char c = dest.charAt(iStart);

            if (c == '\n')
                break;

            if (c != ' ' &&
                    c != '\t') {
                if (!dataBefore) {
                    // indent always after those characters
                    if (c == '{' || c == '+' || c == '-' || c == '*' || c == '/' || c == '%' || c == '^' || c == '=')
                        pt--;

                    dataBefore = true;
                }

                // parenthesis counter
                if (c == '(')
                    --pt;
                else if (c == ')')
                    ++pt;
            }
        }

        // copy indent of this line into the next
        if (iStart > -1) {
            char charAtCursor = dest.charAt(dStart);
            int iEnd;

            for (iEnd = ++iStart;
                 iEnd < dEnd;
                 ++iEnd) {
                char c = dest.charAt(iEnd);

                // auto expand comments
                if (charAtCursor != '\n' &&
                        c == '/' &&
                        iEnd + 1 < dEnd &&
                        dest.charAt(iEnd) == c) {
                    iEnd += 2;
                    break;
                }

                if (c != ' ' &&
                        c != '\t')
                    break;
            }

            indent += dest.subSequence(iStart, iEnd);
        }

        // add new indent
        if (pt < 0)
            indent += "\t";

        // append white space of previous line and new indent
        return source + indent;
    }

    private int getDigitCount() {
        int count = 0;
        int len = getLineCount();
        while (len > 0) {
            count++;
            len /= 10;
        }
        return count;
    }

    @Override
    protected void onDraw(Canvas canvas) {
        int padding = (int) getPixels(getDigitCount() * 10 + 10);
        setPadding(padding, 0, 0, 0);

        int scrollY = getScrollY();
        int firstLine = layout.getLineForVertical(scrollY), lastLine;

        try {
            lastLine = layout.getLineForVertical(scrollY + (getHeight() - getExtendedPaddingTop() - getExtendedPaddingBottom()));
        } catch (NullPointerException npe) {
            lastLine = layout.getLineForVertical(scrollY + (getHeight() - getPaddingTop() - getPaddingBottom()));
        }

        //the y position starts at the baseline of the first line
        int positionY = getBaseline() + (layout.getLineBaseline(firstLine) - layout.getLineBaseline(0));
        drawLineNumber(canvas, layout, positionY, firstLine);
        for (int i = firstLine + 1; i <= lastLine; i++) {
            //get the next y position using the difference between the current and last baseline
            positionY += layout.getLineBaseline(i) - layout.getLineBaseline(i - 1);
            drawLineNumber(canvas, layout, positionY, i);
        }

        super.onDraw(canvas);
    }

    private void drawLineNumber(Canvas canvas, Layout layout, int positionY, int line) {
        int positionX = (int) layout.getLineLeft(line);
        canvas.drawText(String.valueOf(line + 1), positionX + getPixels(2), positionY, paint);
    }

    private float getPixels(int dp) {
        return TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, dp, context.getResources().getDisplayMetrics());
    }

    public interface OnTextChangedListener {
        void onTextChanged(String text);
    }

    /*** Keyboard checking ***/
    @Override
    public boolean onKeyPreIme(int keyCode, KeyEvent event) {
        if (event.getKeyCode() == KeyEvent.KEYCODE_BACK && event.getAction() == KeyEvent.ACTION_UP)
            if (mOnImeBack != null) mOnImeBack.onImeBack();
        return super.dispatchKeyEvent(event);
    }

    public void setBackPressedListener(BackPressedListener listener) {
        mOnImeBack = listener;
    }

    public interface BackPressedListener {
        void onImeBack();
    }
}

File: app/src/main/java/net/theluckycoder/sharpide/widget/CodeEditText.kt
package net.theluckycoder.sharpide.widget

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.os.Handler
import android.preference.PreferenceManager
import android.support.v4.content.ContextCompat
import android.support.v7.widget.AppCompatEditText
import android.text.*
import android.text.style.BackgroundColorSpan
import android.text.style.ForegroundColorSpan
import android.util.AttributeSet
import android.util.Log
import android.util.TypedValue
import android.view.KeyEvent
import net.theluckycoder.sharpide.R
import java.util.regex.Pattern


class CodeEditText : AppCompatEditText {

    companion object {
        private val PATTERN_CLASSES = Pattern.compile(
                "^[\t ]*(Object|Function|Boolean|Symbol|Error|EvalError|InternalError|" +
                        "RangeError|ReferenceError|SyntaxError|TypeError|URIError|" +
                        "Number|Math|Date|String|RegExp|Map|Set|WeakMap|WeakSet|" +
                        "Array|ArrayBuffer|DataView|JSON|Promise|Generator|GeneratorFunction" +
                        "Reflect|Proxy|Intl)\\b",
                Pattern.MULTILINE)
        private val PATTERN_CUSTOM_CLASSES = Pattern.compile(
                "(\\w+[ .])")
        private val PATTERN_KEYWORDS = Pattern.compile(
                "\\b(break|case|catch|class|const|continue|debugger|default|delete|do|yield|" +
                        "else|export|extends|finally|for|function|if|import|in|instanceof|" +
                        "new|return|super|switch|this|throw|try|typeof|var|void|while|with|" +
                        "null|true|false)\\b")
        private val PATTERN_COMMENTS = Pattern.compile("/\\*(?:.|[\\n\\r])*?\\*/|//.*")
        private val PATTERN_SYMBOLS = Pattern.compile("[+\\-*&^!:/|?<>=;,.]")
        private val PATTERN_NUMBERS = Pattern.compile("\\b(\\d*[.]?\\d+)\\b")
    }

    private val mContext: Context
    @Transient private val paint = Paint()
    private var mLayout: Layout? = null
    private val updateHandler = Handler()
    private val onTextChangedListener: OnTextChangedListener? = null
    private val updateDelay = 500
    private var modified = true
    private var colorNumber = 0
    private var colorKeyword = 0
    private var colorClasses = 0
    private var colorComment = 0
    private var colorString = 0
    private val updateRunnable = Runnable {
        val e = text

        onTextChangedListener?.onTextChanged(
                e.toString())

        highlightWithoutChange(e)
    }
    private var mOnImeBack: BackPressedListener? = null

    constructor(context: Context) : super(context) {
        mContext = context
        init(context)
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        mContext = context
        init(context)
    }

    private fun clearSpans(e: Editable) {
        // remove foreground color spans
        run {
            val spans = e.getSpans(0, e.length, ForegroundColorSpan::class.java)

            var n = spans.size
            while (n-- > 0)
                e.removeSpan(spans[n])
        }

        // remove background color spans
        run {
            val spans = e.getSpans(0, e.length, BackgroundColorSpan::class.java)

            var n = spans.size
            while (n-- > 0)
                e.removeSpan(spans[n])
        }
    }

    fun setTextHighlighted(text: CharSequence?) {
        val newText = text ?: ""

        cancelUpdate()

        modified = false
        setText(highlight(SpannableStringBuilder(newText)))
        modified = true

        onTextChangedListener?.onTextChanged(newText.toString())
    }

    private fun init(context: Context) {
        setHorizontallyScrolling(true)

        filters = arrayOf(InputFilter { source, start, end, dest, dStart, dEnd ->
            if (modified &&
                    end - start == 1 &&
                    start < source.length &&
                    dStart < dest.length) {
                val c = source[start]

                if (c == '\n') return@InputFilter autoIndent(source, dest, dStart, dEnd)
            }

            source
        })

        addTextChangedListener(
                object : TextWatcher {

                    override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) = Unit

                    override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) = Unit

                    override fun afterTextChanged(e: Editable) {
                        cancelUpdate()

                        if (!modified) return

                        updateHandler.postDelayed(updateRunnable, updateDelay.toLong())
                    }
                })

        // Set Syntax Colors
        colorNumber = ContextCompat.getColor(context, R.color.syntax_number)
        colorKeyword = ContextCompat.getColor(context, R.color.syntax_keyword)
        colorClasses = ContextCompat.getColor(context, R.color.syntax_class)
        colorComment = ContextCompat.getColor(context, R.color.syntax_comment)
        colorString = ContextCompat.getColor(context, R.color.syntax_string)

        val bgPaint = Paint()
        bgPaint.style = Paint.Style.FILL
        bgPaint.color = Color.parseColor("#eeeeee")

        paint.style = Paint.Style.FILL
        paint.isAntiAlias = true
        paint.color = Color.parseColor("#bbbbbb")
        paint.textSize = getPixels(Integer.parseInt(PreferenceManager.getDefaultSharedPreferences(context).getString("font_size", "16")))
        viewTreeObserver.addOnGlobalLayoutListener { mLayout = layout }
    }

    private fun cancelUpdate() {
        updateHandler.removeCallbacks(updateRunnable)
    }

    private fun highlightWithoutChange(e: Editable) {
        modified = false
        highlight(e)
        modified = true
    }

    private fun highlight(editable: Editable): Editable {
        try {
            // don't use e.clearSpans() because it will
            // remove too much
            clearSpans(editable)

            if (editable.isEmpty()) return editable

            run {
                val m = PATTERN_CLASSES.matcher(editable)
                while (m.find())
                    editable.setSpan(ForegroundColorSpan(colorClasses), m.start(), m.end(), Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
            }

            run {
                val m = PATTERN_CUSTOM_CLASSES.matcher(editable)
                while (m.find())
                    editable.setSpan(ForegroundColorSpan(colorClasses), m.start(), m.end(), Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
            }

            run {
                val m = PATTERN_KEYWORDS.matcher(editable)
                while (m.find())
                    editable.setSpan(ForegroundColorSpan(colorKeyword), m.start(), m.end(), Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
            }

            run {
                val m = PATTERN_SYMBOLS.matcher(editable)
                while (m.find())
                    editable.setSpan(ForegroundColorSpan(colorKeyword), m.start(), m.end(), Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
            }

            run {
                val m = Pattern.compile("\\$\\w+").matcher(editable)
                while (m.find())
                    editable.setSpan(ForegroundColorSpan(colorKeyword), m.start(), m.end(), Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
            }

            run {
                val m = Pattern.compile("\"(.*?)\"|'(.*?)'").matcher(editable)
                while (m.find()) {
                    val spans = editable.getSpans(m.start(), m.end(), ForegroundColorSpan::class.java)
                    for (span in spans)
                        editable.removeSpan(span)
                    editable.setSpan(ForegroundColorSpan(colorString), m.start(), m.end(), Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
                }
            }

            run {
                val m = PATTERN_NUMBERS.matcher(editable)
                while (m.find())
                    editable.setSpan(ForegroundColorSpan(colorNumber), m.start(), m.end(), Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
            }

            val m = PATTERN_COMMENTS.matcher(editable)
            while (m.find()) {
                val spans = editable.getSpans(m.start(), m.end(), ForegroundColorSpan::class.java)
                for (span in spans)
                    editable.removeSpan(span)
                editable.setSpan(ForegroundColorSpan(colorComment), m.start(), m.end(), Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
            }
        } catch (e: IllegalStateException) {
            Log.e("IllegalStateException", e.message, e)
            // raised by Matcher.start()/.end() when
            // no successful match has been made what
            // shouldn't ever happen because of find()
        }

        return editable
    }

    private fun autoIndent(source: CharSequence, dest: Spanned, dStart: Int, dEnd: Int): CharSequence {
        var indent = ""
        var iStart = dStart - 1

        // find start of this line
        var dataBefore = false
        var pt = 0

        while (iStart > -1) {
            val c = dest[iStart]

            if (c == '\n') break

            if (c != ' ' && c != '\t') {
                if (!dataBefore) {
                    // indent always after those characters
                    if (c == '{' || c == '+' || c == '-' || c == '*' || c == '/' || c == '%' || c == '^' || c == '=')
                        pt--

                    dataBefore = true
                }

                // parenthesis counter
                if (c == '(')
                    --pt
                else if (c == ')')
                    ++pt
            }
            --iStart
        }

        // copy indent of this line into the next
        if (iStart > -1) {
            val charAtCursor = dest[dStart]
            var iEnd = ++iStart

            while (iEnd < dEnd) {
                val c = dest[iEnd]

                // auto expand comments
                if (charAtCursor != '\n' &&
                        c == '/' &&
                        iEnd + 1 < dEnd &&
                        dest[iEnd] == c) {
                    iEnd += 2
                    break
                }

                if (c != ' ' && c != '\t')
                    break
                ++iEnd
            }

            indent += dest.subSequence(iStart, iEnd)
        }

        // add new indent
        if (pt < 0) indent += "\t"

        // append white space of previous line and new indent
        return source.toString() + indent
    }

    private val digitCount: Int
        get() {
            var count = 0
            var lineCount = lineCount

            while (lineCount > 0) {
                count++
                lineCount /= 10
            }

            return count
        }

    override fun onDraw(canvas: Canvas) {
        val padding = getPixels(digitCount * 10 + 10).toInt()
        setPadding(padding, 0, 0, 0)

        val scrollY = scrollY
        val firstLine = mLayout!!.getLineForVertical(scrollY)
        val lastLine = try {
            mLayout!!.getLineForVertical(scrollY + (height - extendedPaddingTop - extendedPaddingBottom))
        } catch (e: NullPointerException) {
            mLayout!!.getLineForVertical(scrollY + (height - paddingTop - paddingBottom))
        }

        //the y position starts at the baseline of the first line
        var positionY = baseline + (mLayout!!.getLineBaseline(firstLine) - mLayout!!.getLineBaseline(0))
        drawLineNumber(canvas, mLayout, positionY, firstLine)

        for (i in firstLine + 1..lastLine) {
            //get the next y position using the difference between the current and last baseline
            positionY += mLayout!!.getLineBaseline(i) - mLayout!!.getLineBaseline(i - 1)
            drawLineNumber(canvas, mLayout, positionY, i)
        }

        super.onDraw(canvas)
    }

    private fun drawLineNumber(canvas: Canvas, layout: Layout?, positionY: Int, line: Int) {
        val positionX = layout!!.getLineLeft(line).toInt()
        canvas.drawText((line + 1).toString(), positionX + getPixels(2), positionY.toFloat(), paint)
    }

    private fun getPixels(dp: Int): Float {
        return TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, dp.toFloat(), mContext.resources.displayMetrics)
    }

    interface OnTextChangedListener {
        fun onTextChanged(text: String)
    }

    /*** Keyboard checking  */
    override fun onKeyPreIme(keyCode: Int, event: KeyEvent): Boolean {
        if (event.keyCode == KeyEvent.KEYCODE_BACK && event.action == KeyEvent.ACTION_UP)
            mOnImeBack?.onImeBack()
        return super.dispatchKeyEvent(event)
    }

    interface BackPressedListener {
        fun onImeBack()
    }
}
